Origin: uptream https://github.com/varnishcache/varnish-cache/commit/176f8a075a
Description: Fix VSV00002 Data leak - ‘-sfile’ Stevedore transient objects. Already shipped as part of
Varnish 6.0 and 6.4, it can be safely removed on a 6.x release.
From 6143df703e93f6a599ffcbd8258af7ce45d14576 Mon Sep 17 00:00:00 2001
From: Martin Blix Grydeland <martin@varnish-software.com>
Date: Mon, 18 Sep 2017 16:04:53 +0200
Subject: [PATCH] Avoid buffer read overflow on vcl_error and -sfile

The file stevedore may return a buffer larger than asked for when
requesting storage. Due to lack of check for this condition, the code
to copy the synthetic error memory buffer from vcl_error would overrun
the buffer.

Patch by @shamger

Fixes: #2429
---
 bin/varnishd/cache/cache_fetch.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/bin/varnishd/cache/cache_fetch.c b/bin/varnishd/cache/cache_fetch.c
index 670dc8f..bd3a9b6 100644
--- a/bin/varnishd/cache/cache_fetch.c
+++ b/bin/varnishd/cache/cache_fetch.c
@@ -899,6 +899,8 @@ vbf_stp_error(struct worker *wrk, struct busyobj *bo)
 		l = ll;
 		if (VFP_GetStorage(bo->vfc, &l, &ptr) != VFP_OK)
 			break;
+		if (l > ll)
+			l = ll;
 		memcpy(ptr, VSB_data(synth_body) + o, l);
 		VFP_Extend(bo->vfc, l);
 		ll -= l;
-- 
2.1.4

